### DISCUSSION TITLE
Extending type-checking to general chunk assertions

### DISCUSSION THREAD

**Discussion by balacij** [JacquesCarette/Drasil#3134] [Ideas]
Related to #3093 and #2873 (specifically, @JacquesCarette 's comment about adding a `State` for gradual chunk registration in the chunk database)

With #3093, we add a typeclass, `RequiresChecking`, which exposes a series of expressions (of a specific type and w.r.t. a specific type universe) that need to be type-checked. If we desired, we can upgrade this assertion functionality to the level of chunk creation and registration in the chunk database. We can also make the type-checking assertions more abstract, making them a specific variant of general "general required assertions", and use the same code as well. With this, if/when we have Drasil interpreting external code, we would be able to assert various properties about the chunks.

In some sense, this would be creating LiquidHaskell-like constraints for our chunk types.

_aside: why does this relate to #2873?_ Without #2873, there is extra complexity associated with implementing this. Specifically, we would need to form and "check" each chunk in a precise order (order found separately). Doing it on-the-fly is just easier, but not necessarily required.

**Comment by balacij** [comment]
For example, asides from type-checking expressions, we might want to create assertions for documentation (we might be able to plug in an external grammar checker to add warnings and errors).

Any chunk that isn't necessarily "correct by construction", or where I've previously asked if we want Dependent Types, (e.g., any chunk that require smart constructors to assert certain properties about them) might find this to be helpful.

**Comment by balacij** [comment]
One of the "nice" things about an idea like this is that it forces us to categorize assertions, and we will also be able to ignore (or otherwise) assertions by categories. It _might_ also mean we can also remove many of our smart constructors that exist purely for asserting things. For the smart constructors that exist for other reasons, we would be able to learn more about what they do and why the smart constructors are important for Drasil/chunk creation. Furthermore, we would have more implicit-data about our chunk types made explicit.

**Reply by JacquesCarette** [reply]
Yes, I agree, this is a very nice idea.

**Reply by smiths** [reply]
Yes, I like it too.

**Reply by balacij** [reply]
:smile: I'm glad! Hopefully we can return to this idea some time post- #2873 then. I would love to implement something like this, it fits in well with the gradual chunk registration, I think.
